src/hashrate.rs

```diff
@@ -0,0 +1,74 @@
+use std::time::{Duration, Instant};
+use colored::*;
+
+
+pub fn get_unit(hashrate: f32) -> (f32, String) {
+    let mut hashrate = hashrate as f32;
+    let mut unit = "";
+    let units = vec!["k", "M", "G", "T", "P", "E", "Z", "Y"];
+    for u in units.iter() {
+        if hashrate > 1000.0 {
+            hashrate /= 1000.0;
+            unit = u;
+        } else {
+            break;
+        }
+    }
+    (hashrate, format!("{}H/s", unit))
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Hashrate {
+    pub worker_id: u32,
+    value: f32,
+    counter: u32,
+    counter_max: u32,
+    sampling_start: Instant,
+    report_start: Instant,
+    report_interval: Duration,
+}
+
+impl std::fmt::Display for Hashrate {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (value, unit) = get_unit(self.value);
+        write!(f, "Worker {} Hashrate = {} {}",
+               format!("#{}", self.worker_id).yellow(),
+               format!("{:.3}", value).red(),
+               unit)
+    }
+}
+
+impl core::iter::Sum for Hashrate {
+    fn sum<I: Iterator<Item=Hashrate>>(iter: I) -> Hashrate {
+        iter.fold(Hashrate::new(0, 0.0), |a, b| Hashrate::new(0, a.value + b.value))
+    }
+}
+
+impl Hashrate {
+    pub fn new(worker_id: u32, value: f32) -> Self {
+        Self { worker_id, value, counter: 0, 
+               counter_max: 32,
```

Comment:
32 is hardcoded. Define this in as a constant at the beginning of the file
const HASHRATE_COUNTER_MAX: usize = 32;
and then: counter_max: HASHRATE_COUNTER_MAX

Reply:
`usize` or `u32`? which one is better?

Reply:
usize. Better also use a higher counter_max imo

Reply:
I purposely chosen a small `HASHRATE_COUNTER_MAX=32` because if someone's machine is not powerful, it may not be able to calculate hashrate. If you have a better suggestion, I can change it.